src/wms/src/main/java/org/geoserver/wms/legendgraphic/JSONLegendGraphicBuilder.java

```diff
@@ -0,0 +1,843 @@
+
+            Expression opac = entry.getOpacity();
+            if (opac != null) {
+                ent.element(OPACITY, printExpression(opac, Double.class));
+            }
+            entries.add(ent);
+        }
+        JSONObject cm = new JSONObject();
+        if (entries.size() > 0) {
+            cm.element(ENTRIES, entries);
+            int type = colorMap.getType();
+            switch (type) {
+                case ColorMap.TYPE_INTERVALS:
+                    cm.element(COLORMAP_TYPE, "intervals");
+                    break;
+                case ColorMap.TYPE_RAMP:
+                    cm.element(COLORMAP_TYPE, "ramp");
+                    break;
+                case ColorMap.TYPE_VALUES:
+                    cm.element(COLORMAP_TYPE, "values");
+                    break;
+            }
+
+            ret.element(COLORMAP, cm);
+        }
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param fill
+     * @return
+     */
+    private JSONObject processFill(JSONObject ret, Fill fill) {
+        if (fill == null) {
+            return ret;
+        }
+        boolean filled = false;
+
+        if (fill.getGraphicFill() != null) {
+            filled = true;
+            JSONObject jGraphic = new JSONObject();
+            jGraphic = processGraphic(jGraphic, fill.getGraphicFill());
+            ret.element(GRAPHIC_FILL, jGraphic);
+        }
+        if (!filled) {
+            ret.element(FILL, printExpression(fill.getColor(), Color.class));
+            ret.element(FILL_OPACITY, printExpression(fill.getOpacity(), Number.class));
+        }
+        return ret;
+    }
+
+    /**
+     * @param ret
+     * @param graphic
+     * @return
+     */
+    private JSONObject processGraphic(JSONObject ret, Graphic graphic) {
+
+        Catalog catalog = wms.getCatalog();
+        StyleInfo styleByName = catalog.getStyleByName(styleName);
+        String wsName = null;
+        if (styleByName != null) {
+            WorkspaceInfo ws = styleByName.getWorkspace();
+            if (ws != null) wsName = ws.getName();
+        }
+        ret.element(
+                "url",
+                IconPropertyExtractor.extractProperties(miniStyle, (SimpleFeature) feature)
+                        .href(baseURL, wsName, styleName));
+        JSONArray jGraphics = new JSONArray();
+        List<GraphicalSymbol> gSymbols = graphic.graphicalSymbols();
+        for (GraphicalSymbol g : gSymbols) {
+            JSONObject jGraphic = processGraphicalSymbol(g);
+
+            jGraphics.add(jGraphic);
+        }
+
+        Expression size2 = graphic.getSize();
+        if (size2 != null) {
+            ret.element(SIZE, printExpression(size2, Number.class));
+        }
+        Expression opacity = graphic.getOpacity();
+        if (opacity != null) {
+            ret.element(OPACITY, printExpression(opacity, Number.class));
+        }
+        Expression rotation = graphic.getRotation();
```

Comment:
I'd add displacement and anchor encoding too, they can be specified in SLD 1.1 and the renderer actually honors them.